ActivatePubKey/handler.ts

```diff
@@ -0,0 +1,215 @@
+>;
+export const ActivatePubKeyHandler = (
+  popDocumentReader: PopDocumentReader,
+  popDocumentWriter: PopDocumentWriter,
+  assertionWriter: AssertionWriter
+): ActivatePubKeyHandler => (
+  context,
+  assertion_ref,
+  body
+): ReturnType<ActivatePubKeyHandler> =>
+  pipe(
+    popDocumentReader(assertion_ref),
+    TE.mapLeft(domainErrorToResponseError),
+    TE.filterOrElseW(isPendingLollipopPubKey, doc => {
+      const err = `Unexpected status on pop document during activation: ${doc.status}`;
+      context.log.error(err);
+      return ResponseErrorInternal(err);
+    }),
+    TE.bindTo("popDocument"),
+    TE.bindW("assertionFileName", () =>
+      pipe(
+        `${body.fiscal_code}-${assertion_ref}`,
+        AssertionFileName.decode,
+        TE.fromEither,
+        TE.mapLeft(errors => {
+          const err = `Could not decode assertionFileName | ${readableReportSimplified(
+            errors
+          )}`;
+          context.log.error(err);
+          return ResponseErrorInternal(err);
+        }),
+        TE.chainFirst(assertionFileName =>
+          pipe(
+            assertionWriter(assertionFileName, body.assertion),
+            TE.mapLeft(error => {
+              const err = error.detail;
+              context.log.error(err);
+              return ResponseErrorInternal(error.detail);
+            })
+          )
+        )
+      )
+    ),
+    TE.bindW("jwkPubKeyFromString", ({ popDocument }) =>
+      pipe(
+        popDocument.pubKey,
+        JwkPublicKeyFromToken.decode,
+        TE.fromEither,
+        TE.mapLeft(errors => {
+          const err = `Could not decode public key | ${readableReportSimplified(
+            errors
+          )}`;
+          context.log.error(err);
+          return ResponseErrorInternal(err);
+        })
+      )
+    ),
+    TE.bindW("assertionRefs", ({ popDocument, jwkPubKeyFromString }) =>
+      pipe(
+        getAllAssertionsRef(
+          MASTER_HASH_ALGO,
+          getAlgoFromAssertionRef(popDocument.assertionRef),
+          jwkPubKeyFromString
+        ),
+        TE.mapLeft((error: Error) => {
+          const err = error.message;
+          context.log.error(err);
+          return ResponseErrorInternal(err);
+        })
+      )
+    ),
+    TE.bindW(
+      "retrievedPopDocument",
+      ({ popDocument, assertionRefs, assertionFileName }) =>
+        activatePubKeyForAssertionRef(popDocumentWriter, context)(
+          assertionFileName,
+          assertionRefs.master,
+          body,
+          popDocument.pubKey
+        )
+    ),
+    TE.chain(
+      ({
+        assertionRefs,
+        assertionFileName,
+        popDocument,
+        retrievedPopDocument
+      }) =>
+        assertionRefs.used
+          ? activatePubKeyForAssertionRef(popDocumentWriter, context)(
+              assertionFileName,
+              assertionRefs.used,
+              body,
+              popDocument.pubKey
+            )
+          : TE.of(retrievedPopDocument)
+    ),
```

Comment:
why not sequencing TaskEither here?

Reply:
do you mean replacing ternary operator with a fromPredicate flow?

Reply:
Nope , I mean sequencing `activatePubKeyForAssertionRef` but now I see that you use `retrievedPopDocument` from the first one to the second one.